ui/pages/permissions-connect/connect-page/utils.ts

```diff
@@ -70,3 +77,177 @@ export function getCaip25PermissionsResponse(
+  requestedCaip25CaveatValue: Caip25CaveatValue,
+) {
+  const requiredAccounts = Object.values(
+    requestedCaip25CaveatValue.requiredScopes,
+  ).flatMap((scope) => scope.accounts);
+
+  const optionalAccounts = Object.values(
+    requestedCaip25CaveatValue.optionalScopes,
+  ).flatMap((scope) => scope.accounts);
+
+  return new Set([...requiredAccounts, ...optionalAccounts]);
+}
+
+/**
+ * Extracts chain IDs from both required and optional scopes in CAIP-25 caveat value.
+ * Converts EVM chain IDs to hexadecimal format and keeps other chain identifiers as is.
+ *
+ * @param caveatValue - The CAIP-25 caveat value containing scopes.
+ * @returns Array of chain IDs.
+ */
+export function getRequestedChainIds(caveatValue: Caip25CaveatValue): string[] {
+  const allScopes = [
+    ...Object.keys(caveatValue.requiredScopes),
+    ...Object.keys(caveatValue.optionalScopes),
+  ];
+
+  return [...new Set(allScopes)].map((scope) => {
+    if (scope.startsWith('eip155:')) {
+      const chainId = scope.split(':')[1];
+      return `0x${parseInt(chainId, 10).toString(16)}`;
+    }
+
+    return scope;
+  });
+}
+
+export function getAllRequestedChainIds(
+  requestedCaip25CaveatValue: Caip25CaveatValue,
+) {
+  const allScopes = [
+    ...Object.keys(requestedCaip25CaveatValue.requiredScopes),
+    ...Object.keys(requestedCaip25CaveatValue.optionalScopes),
+  ];
+
+  return [...new Set(allScopes)];
+}
+
+/**
+ * Filters networks based on the CAIP request scopes.
+ *
+ * @param networks - Network configurations.
+ * @param requestedCaip25CaveatValue - CAIP-25 caveat value.
+ * @returns Filtered network configurations matching the requested scopes.
+ */
+export function getFilteredNetworks(
+  networks: Record<string, NetworkConfiguration>,
+  requestedCaip25CaveatValue: Caip25CaveatValue,
+): Record<string, NetworkConfiguration> {
+  const filteredNetworks: Record<string, NetworkConfiguration> = {};
+  const allScopes = {
+    ...(requestedCaip25CaveatValue.requiredScopes || {}),
+    ...(requestedCaip25CaveatValue.optionalScopes || {}),
+  };
+
+  const hasEipRequest = Object.keys(allScopes).some(
+    (scope) => scope.startsWith('eip155:') || scope.startsWith('wallet:eip155'),
+  );
+
+  Object.entries(networks).forEach(([chainId, network]) => {
+    if (chainId.startsWith('0x')) {
+      if (hasEipRequest) {
+        filteredNetworks[chainId] = network;
+      }
+    } else if (Object.keys(allScopes).includes(chainId)) {
+      filteredNetworks[chainId] = network;
+    }
+  });
+
+  return filteredNetworks;
+}
+
+/**
+ * Gets the default accounts for the requested namespaces.
+ * We need at least one default per requested namespace
+ * if there are more explicitly requested accounts, use those instead
+ * for that namespace
+ *
+ * @param requestedNamespaces - The namespaces requested.
+ * @param supportedRequestedAccounts - The supported requested accounts.
+ * @param allAccounts - All available accounts.
+ */
+export function getDefaultAccounts(
```

Comment:
this helper needs to also consider the currently selected account, and we need specs